### PR TITLE
[wemo] add annotations and remove usage of apache.commons.*

### PR DESCRIPTION
**Pull request by hmerk**
Related to #7722

Signed-off-by: Hans-Jörg Merk <github@hmerk.de>

### CODE REVIEW

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoLightHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoLightHandler.java:258]: You should cache `wemoHttpCaller` to a local variable and perform your logic on that instead. That should fix the null checker warnings.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java:31]: Why not have this get passed in through the constructor so that you don't have to make it nullable?

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (cpmeister) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:210]: Isn't this changing the behavior? Why are you changing this?

**Review by cpmeister — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoLightHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoLightHandler.java:258]: Changed in all handlers

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java:31]: Sorry, don't understand you, can you give an example please.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java
Comment (J-N-K) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java:55]: Wouldn't it be better to use unbescape (like in #9835) instead of programming an escape/unescape method?

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:210]: No, I don't think so. subscriptionState is defined as boolean and only set true or false in the binding. So in my opinion, this check was superflient. Don't know whay it was added.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java:55]: Did this before I saw you used unbescape. We could change it again, but what would be the benefit ?

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java:31]:
```
public abstract class AbstractWemoHandler extends BaseThingHandler {

    protected WemoHttpCall wemoHttpCaller;

    public AbstractWemoHandler(Thing thing, WemoHttpCall wemoHttpCaller) {
        super(thing);
        this.wemoHttpCaller = wemoHttpCaller;
    }

```

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (cpmeister) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:208]: Are the `subscriptionState` values ever used anywhere? If not then you should make `subscriptionState` a Set instead of a Map.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/AbstractWemoHandler.java:31]: Done

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:208]: The values are used in onServiceSubscribed(), onSubscription() and removeSubscription()

**Comment by hmerk** [comment]
@cpmeister How do we proceed ??

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java
Comment (J-N-K) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/WemoUtil.java:55]: I'm fine with your solution (that's why I just commented and didn't request a change).The only benefit would be that we don't need to take care for different implementations of the same functionality.

**Comment by hmerk** [comment]
Hm, build failure seems to be unrelated to my PR, even tests did not fail...

**Comment by wborn** [comment]
I see a failing test:

```
TEST org.openhab.binding.wemo.internal.handler.test.WemoHandlerOSGiTest#assertThatThingHandlesOnOffCommandCorrectly() <<< ERROR:
Wanted but not invoked:
wemoHttpCall.executeCall(
    <any>,
    <any>,
    <Capturing argument>
);
-> at org.openhab.binding.wemo.internal.handler.test.WemoHandlerOSGiTest.assertThatThingHandlesOnOffCommandCorrectly(WemoHandlerOSGiTest.java:95)
Actually, there were zero interactions with this mock.

Wanted but not invoked:
wemoHttpCall.executeCall(
    <any>,
    <any>,
    <Capturing argument>
);
-> at org.openhab.binding.wemo.internal.handler.test.WemoHandlerOSGiTest.assertThatThingHandlesOnOffCommandCorrectly(WemoHandlerOSGiTest.java:95)
Actually, there were zero interactions with this mock.

        at org.openhab.binding.wemo.internal.handler.test.WemoHandlerOSGiTest.assertThatThingHandlesOnOffCommandCorrectly(WemoHandlerOSGiTest.java:95)
```

**Comment by hmerk** [comment]
@wborn Thanks for pointing out. Strage though, Jenkins maked the test „green“
Don‘t know if I will be able to fix this, as I did not write the tests and don‘t have a clue atm.
Any hint, how I can run the test locally in my dev environment ?

**Comment by wborn** [comment]
Just right click the `itest.bndrun` file and choose `Run As -> Bnd OSGi Test Launcher (JUnit)`. That will start and open the JUnit test view with those integration tests:

itest

**Comment by hmerk** [comment]
Thanks @wborn 👍

**Comment by wborn** [comment]
Hi @hmerk! I was able to fix the itests on your branch with the changes in this commit https://github.com/wborn/openhab-addons/commit/4fd1b28ac7d26df792342bb194af4daebfd22a2b.

I added a `WemoHttpCallFactory` interface that is used by the `WemoHandlerFactory` when available, so the itests can use a `WemoHttpCall` mock during the tests.

Another issue was that the `servicePing` always returns `false` when running the itest. So I added a `serviceAvailableFunction` in `WemoUtil` that is replaced with a function that always returns `true` while running the itests.

**Comment by hmerk** [comment]
That's really great news, thanks @wborn !

**Comment by hmerk** [comment]
Hmm, seems I screewed up my branch making Jenkins unhappy.
Going to save all the work putting into this PR and submit a new clean one, then closing this PR

**Comment by wborn** [comment]
If you are happy with my changes, just create a PR from my branch. :wink:

**Comment by wborn** [comment]
I can also force push it over your branch here. The only difference with your current branch is that I squashed your commits so I could more easily resolve the merge conflicts I introduced with the commons-lang PR.

**Comment by hmerk** [comment]
Thanks @wborn, but I made some mistakes with the merge conflicts, that I am resolving now. Seems that I am nearly done without needing to close this PR.
Just have to check if I cherry picked your commit the right way.. ;-)

**Comment by hmerk** [comment]
@wborn could you please check this again, still failing itests:
TEST org.openhab.binding.wemo.internal.handler.test.WemoLightHandlerOSGiTest#handleREFRESHCommandForChannelSTATE() <<< ERROR: 
Wanted but not invoked:
wemoHttpCall.executeCall(
    <any>,
    <any>,
    <Capturing argument>
);
-> at org.openhab.binding.wemo.internal.handler.test.WemoLightHandlerOSGiTest.assertRequestForCommand(WemoLightHandlerOSGiTest.java:177)
Actually, there were zero interactions with this mock.
Wanted but not invoked:
wemoHttpCall.executeCall(
    <any>,
    <any>,
    <Capturing argument>
);

**Comment by wborn** [comment]
The main issue was this [missing change](https://github.com/hmerk/openhab-addons/pull/1/files#diff-65ab7e9813d964d69258665460272f07bd95d36f2dd460bf666e7e9d32911665R104), if you merge https://github.com/hmerk/openhab-addons/pull/1 it should work.

**Comment by hmerk** [comment]
Thanks @wborn, Jenkins is happy again, we are all good for a merge I think....
@cpmeister, waiting for your approval ....

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:299]: What is the reason for this change?

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java:208]: You can skip these null checks by calling the method on non-null objects `"0".equals(binaryState)`

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java:561]: You can also use `Instant.ofEpochSecond` so you don't need this conversion.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java
Comment (wborn) [inline bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java:81]: You can reduce the code by using a lambda instead of a Runnable:

```java
() -> {
    /* code that is in the run method of the Runnable */
};

**Review by wborn — COMMENTED** [review]
Great! The code looks nice now. :-)
I saw a few small improvements:

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:412]: Done

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java:561]: Done

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoDimmerHandler.java:208]: Done

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoHandler.java:81]: Would prefer to leave it as si for now, cause never used a lambda here.

File: bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java
Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:299]: Introduced with this Security advisory:
https://github.com/openhab/openhab-addons/security/advisories/GHSA-r2hc-pmr7-4c9r

Comment (wborn) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java:299]: My point is that you are removing the fix of https://github.com/openhab/openhab-addons/commit/b0a15b48a3e5c07e4657514c6841cafcaa258175#diff-4780ada0429e132073d2ae9d5605f6fe53a3f398fff40ea05a3869e15b7b2427. I.e. you reintroduce the vulnerability.

Comment (hmerk) [reply bundles/org.openhab.binding.wemo/src/main/java/org/openhab/binding/wemo/internal/handler/WemoCoffeeHandler.java]: @wborn readded ...

**Review by wborn — APPROVED** [review]
Thanks @hmerk, the code now LGTM! Perhaps another @openhab/add-ons-maintainers can have a quick second look. This PR also contains some of my own code. :wink:

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks for the massive cleanup!

**Pull request merged**
